### PR TITLE
Add tips detection to auto mapping in bone mapper

### LINKED ISSUE #89725: Skeleton Retarget for ReadyPlayerMe avatars picks incorrect bones
### Tested versions

4.3-dev5

### System information

Windows 11, Vulkan Mobile, NVidia RTX 3070 TI

### Issue description

ReadyPlayerMe produce customized avatars for game players. Importing these avatars using the Skeleton Retarget system results in the wrong finger bones being selected and renamed. The user can correct this by manually picking every bone in the left and right hands; however due to the popularity of ReadyPlayerMe avatars it would be preferrable if these could be identified and handled correctly with minimal effort.

### Steps to reproduce

This Zip file [65fa409029044c117cbd3e3c.zip](https://github.com/godotengine/godot/files/14676308/65fa409029044c117cbd3e3c.zip) contains a ReadyPlayerMe avatar.
image

Importing through the standard Skeleton Retarget method results in the following finger-bone issues:

| Actual Bone Type | ReadyPlayerMe Name | Name after Skeleton Retarget |
| :---- | :---- | :---- |
| Left Thumb Metacarpal | LeftHandThumb1 | LeftHandThumb1 |
| Left Thumb Proximal | LeftHandThumb2 | LeftThumbMetacarpal |
| Left Thumb Distal | LeftHandThumb3 | LeftThumbProximal |
| Left Thumb Tip | LeftHandThumb4 |  LeftThumbDistal |
| Left Index Proximal | LeftHandIndex1 | LeftHandIndex1 |
| Left Index Intermediate | LeftHandIndex2 | LeftIndexProximal |
| Left Index Distal | LeftHandIndex3 | LeftIndexIntermediate |
| Left Index Tip | LeftHandIndex4 | LeftIndexDistal |
| Left Middle Proximal | LeftHandMiddle1 | LeftHandMiddle1 |
| Left Middle Intermediate | LeftHandMiddle2 | LeftMiddleProximal |
| Left Middle Distal | LeftHandMiddle3 | LeftMiddleIntermediate |
| Left Middle Tip | LeftHandMiddle4 | LeftMiddleDistal |
| Left Ring Proximal | LeftHandRing1 | LeftHandRing1 |
| Left Ring Intermediate | LeftHandRing2 | LeftRingProximal |
| Left Ring Distal | LeftHandRing3 | LeftRingIntermediate |
| Left Ring Tip | LeftHandRing4 | LeftRingDistal |
| Left Pinky Proximal | LeftHandPinky1 | LeftHandPinky1 |
| Left Pinky Intermediate | LeftHandPinky2 | LeftLittleProximal |
| Left Pinky Distal | LeftHandPinky3 | LeftLittleIntermidate |
| Left Pinky Tip | LeftHandPinky4 | LeftLittleDistal |

### Minimal reproduction project (MRP)

EDIT(lyuma): Added some more test files which do not contain any mesh. The mesh vertex-based approach is not sufficient to solve these files:
[simple_vrm_rig.glb.gz](https://github.com/godotengine/godot/files/14814474/simple_vrm_rig.glb.gz) - This one is in T-pose
[simple_vrm_rig_anim4.blend](https://github.com/godotengine/godot/files/14814381/simple_vrm_rig_anim4.blend.pptx)
[simple_vrm_rig_nlaonly2.fbx](https://github.com/godotengine/godot/files/14814384/simple_vrm_rig_nlaonly2.fbx.pptx) - this is animated and not in t-pose

### PR DESCRIPTION
**Pull request by TokageItLab**
Fixes #89725.

GodotHumanoid does not have a metacarpal for fingers other than the thumb. To avoid assigning them incorrectly, it is necessary to assign from the ends. However, some models might have "Tips" or "Leaf" bones to provide length information, so assigning from the ends is problematic.

This PR, make auto mapping can detected Tips and the mapping will be offset when candidate bones for the fingers has the name Tip or Leaf, or if none of the meshes have weights bound to them.

~If the joint count of fingers excluding the thumb is the same, if even one Tip is detected, it will be apply by all bones. This is because **the ReadyPlayerMe model is weird since only the index finger tips has weight**.~

~This becomes a problem in cases where you intentionally have tips on only one finger, so if the number of bones is different, determine the tips individually. Still it may be a problem if the number of fingers is the same and you forget to bind only a specific finger, but I consider it is a rare case.~

Make the detection based on the name match and the number of children in the thumb method by @lyuma.

### CODE REVIEW

**Comment by lyuma** [comment]
Added some more test files which do not contain any mesh.

I am skeptical of using mesh vertex-based approaches, since files such as these do not contain any mesh at all:
[simple_vrm_rig.glb.gz](https://github.com/godotengine/godot/files/14814474/simple_vrm_rig.glb.gz) - This one is in T-pose
[simple_vrm_rig_anim4.blend](https://github.com/godotengine/godot/files/14814381/simple_vrm_rig_anim4.blend.pptx)
[simple_vrm_rig_nlaonly2.fbx](https://github.com/godotengine/godot/files/14814384/simple_vrm_rig_nlaonly2.fbx.pptx) - this is animated and not in t-pose

I'll test the PR and see if I can offer any changes.
For example, I ran into a similar problem in Unidot (which uses a copy of Godot's skeleton profile guesser) and I came up with this approach:
https://github.com/V-Sekai/unidot_importer/commit/4f4f62ca886b054e28ff48e540be3f3cbf043e28
However, I am a little concerned that this does not sufficiently address cases where there are two _End bones (which happens when importing and re-exporting an animation FBX with the default setting of "Add Leaf Bone"
But maybe it will mitigate the most common cases

**Comment by TokageItLab** [comment]
Indeed it is little skeptical, but since BoneMap is reusable, so I think this implementation would be a not big problem. So it would work as a workaround to use a BoneMap generated for a model with a mesh in an animation file without a mesh, as long as they have the same bone name.

**Comment by TokageItLab** [comment]
From discussed with @lyuma, we think we can add a couple of decision points, so I will add them.

So the steps with the following priority order should be used for estimation:

1. put "distal" in the search suggestions
2. if the number of thumb bones is 4, there are Tips
3. look at the mesh weights

**Comment by TokageItLab** [comment]
Detection by the number of thumb children conflicts with detection from weights. Probably, given the existence of skeletons without skins, detection by the number of thumb children should be preferred. Therefore, I will adopt @lyuma's method overall, thanks.

File: editor/plugins/bone_map_editor_plugin.h [hunk 1]
```diff
@@ -45,6 +45,7 @@
 #include "scene/gui/color_rect.h"
 #include "scene/gui/dialogs.h"
 #include "scene/resources/bone_map.h"
+#include "scene/resources/surface_tool.h"
```
Comment (lyuma) [inline editor/plugins/bone_map_editor_plugin.h:48]: unused import

File: editor/plugins/bone_map_editor_plugin.h [hunk 2]
```diff
@@ -182,6 +183,8 @@ class BoneMapper : public VBoxContainer {
 	bool is_match_with_bone_name(const String &p_bone_name, const String &p_word);
 	int search_bone_by_name(Skeleton3D *p_skeleton, const Vector<String> &p_picklist, BoneSegregation p_segregation = BONE_SEGREGATION_NONE, int p_parent = -1, int p_child = -1, int p_children_count = -1);
 	BoneSegregation guess_bone_segregation(const String &p_bone_name);
+	LocalVector<LocalVector<SurfaceTool::Vertex>> get_vertices_list(Skeleton3D *p_skeleton);
+	bool is_bound_by_surface(LocalVector<LocalVector<SurfaceTool::Vertex>> &p_vertices_list, Skeleton3D *p_skeleton, int p_bone_idx);
```
Comment (lyuma) [inline editor/plugins/bone_map_editor_plugin.h:187]: unused declarations

**Review by lyuma — APPROVED** [review]
Reviewed on a call with fire. I think this is good. It can use more testing, but that's what beta is for.

What's important is this improves Mixamo and ReadyPlayerMe compatibility by adding a few heuristics, so I think it's fairly safe.

**Review by fire — APPROVED** [review]

File: editor/plugins/bone_map_editor_plugin.cpp [hunk 3]
```diff
@@ -894,6 +896,33 @@ void BoneMapper::auto_mapping_process(Ref<BoneMap> &p_bone_map) {
 					search_path.push_back(finger);
 					finger = skeleton->get_bone_parent(finger);
 				}
+				// Tips detection by name matching with "distal" from root.
+				for (int j = search_path.size() - 1; j >= 0; j--) {
+					if (!RegEx("distal").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_null()) {
```
Comment (AThousandShips) [inline editor/plugins/bone_map_editor_plugin.cpp:901]:
```suggestion
					if (RegEx("distal").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_valid()) {
```

File: editor/plugins/bone_map_editor_plugin.cpp [hunk 4]
```diff
@@ -894,6 +896,33 @@ void BoneMapper::auto_mapping_process(Ref<BoneMap> &p_bone_map) {
...
+						tips_index = j - 1;
+						break;
+					}
+				}
+				// Tips detection by name matching with "tip|leaf" from end.
+				if (tips_index < 0) {
+					for (int j = 0; j < search_path.size(); j++) {
+						if (!RegEx("tip|leaf").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_null()) {
```
Comment (AThousandShips) [inline editor/plugins/bone_map_editor_plugin.cpp:909]:
```suggestion
						if (RegEx("tip|leaf").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_valid()) {
```

File: editor/plugins/bone_map_editor_plugin.cpp [hunk 5]
```diff
@@ -985,6 +1025,33 @@ void BoneMapper::auto_mapping_process(Ref<BoneMap> &p_bone_map) {
 					search_path.push_back(finger);
 					finger = skeleton->get_bone_parent(finger);
 				}
+				// Tips detection by name matching with "distal" from root.
+				for (int j = search_path.size() - 1; j >= 0; j--) {
+					if (!RegEx("distal").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_null()) {
```
Comment (AThousandShips) [inline editor/plugins/bone_map_editor_plugin.cpp:1030]:
```suggestion
					if (RegEx("distal").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_valid()) {
```

File: editor/plugins/bone_map_editor_plugin.cpp [hunk 6]
```diff
@@ -985,6 +1025,33 @@ void BoneMapper::auto_mapping_process(Ref<BoneMap> &p_bone_map) {
...
+						tips_index = j - 1;
+						break;
+					}
+				}
+				// Tips detection by name matching with "tip|leaf" from end.
+				if (tips_index < 0) {
+					for (int j = 0; j < search_path.size(); j++) {
+						if (!RegEx("tip|leaf").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_null()) {
```
Comment (AThousandShips) [inline editor/plugins/bone_map_editor_plugin.cpp:1038]:
```suggestion
						if (RegEx("tip|leaf").search(skeleton->get_bone_name(search_path[j]).to_lower()).is_valid()) {
```

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**